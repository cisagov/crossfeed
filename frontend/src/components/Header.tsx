import React, { useState } from 'react';
import clsx from 'classnames';
import { NavLink, Link } from 'react-router-dom';
import { makeStyles } from '@material-ui/core/styles';
import {
  AppBar,
  Toolbar,
  IconButton,
  Drawer,
  ListItem,
  List
} from '@material-ui/core';
import {
  Menu as MenuIcon,
  AccountCircle as UserIcon
} from '@material-ui/icons';
import { useAuthContext } from 'context';
import logo from '../assets/cisa_logo.png';

const GLOBAL_ADMIN = 4;
const ORG_ADMIN = 2;
const ORG_USER = 1;
const ALL_USERS = GLOBAL_ADMIN | ORG_ADMIN | ORG_USER;

const useStyles = makeStyles((theme) => ({
  inner: {
    maxWidth: 1440,
    width: '100%',
    margin: '0 auto'
  },
  menuButton: {
    marginRight: theme.spacing(2),
    display: 'block',
    [theme.breakpoints.up('lg')]: {
      display: 'none'
    }
  },
  logo: {
    display: 'none',
    height: 60,
    padding: theme.spacing(),
    paddingLeft: 0,
    [theme.breakpoints.up('sm')]: {
      display: 'block'
    }
  },
  spacing: {
    flexGrow: 1
  },
  activeLink: {
    '&:after': {
      content: "''",
      position: 'absolute',
      bottom: 0,
      left: 0,
      width: '100%',
      height: 2,
      backgroundColor: 'white'
    }
  },
  activeMobileLink: {
    fontWeight: 700,
    '&:after': {
      content: "''",
      position: 'absolute',
      top: 0,
      bottom: 0,
      left: 0,
      height: '100%',
      width: 2,
      backgroundColor: theme.palette.primary.main
    }
  },
  link: {
    position: 'relative',
    color: 'white',
    textDecoration: 'none',
    margin: `0 ${theme.spacing()}px`,
    padding: theme.spacing(),
    borderBottom: '2px solid transparent',
    fontWeight: 600
  },
  userLink: {
    display: 'flex',
    alignItems: 'center',

    '& svg': {
      marginRight: theme.spacing()
    }
  },
  mdNav: {
    display: 'none',
    [theme.breakpoints.up('md')]: {
      display: 'block'
    }
  },
  lgNav: {
    display: 'none',
    [theme.breakpoints.up('lg')]: {
      display: 'block'
    }
  },
  mobileNav: {
    padding: `${theme.spacing(2)}px ${theme.spacing()}px`
  }
}));

interface NavItem {
  title: string;
  path: string;
  users: number;
}

export const Header: React.FC = () => {
  const classes = useStyles();
  const { currentOrganization, user } = useAuthContext();
  const [navOpen, setNavOpen] = useState(false);

  let userLevel = 0;
  if (user && user.isRegistered) {
    if (user.userType === 'standard') {
      if (currentOrganization?.userIsAdmin) {
        userLevel = ORG_ADMIN;
      } else {
        userLevel = ORG_USER;
      }
    } else {
      userLevel = GLOBAL_ADMIN;
    }
  }

  const navItems: NavItem[] = [
    { title: 'Inventory', path: '/', users: ALL_USERS },
    { title: 'Search', path: '/search', users: ALL_USERS },
    { title: 'Vulnerabilities', path: '/vulnerabilities', users: ALL_USERS },
    { title: 'Risk Summary', path: '/risk', users: ALL_USERS },
    {
      title: 'Organization Settings',
      path: '/organization',
      users: ORG_ADMIN | GLOBAL_ADMIN
    },
    {
      title: 'My Organizations',
      path: '/organizations',
      users: ORG_USER | ORG_ADMIN
    },
    {
      title: 'Manage Organizations',
      path: '/organizations',
      users: GLOBAL_ADMIN
    },

    { title: 'Scans', path: '/scans', users: GLOBAL_ADMIN },
    { title: 'Manage Users', path: '/users', users: GLOBAL_ADMIN }
  ].filter(({ users }) => (users & userLevel) > 0);

  const desktopNavItems: JSX.Element[] = navItems.map(({ title, path }) => (
    <NavLink
      to={path}
      key={title}
      activeClassName={classes.activeLink}
      className={classes.link}
      exact={true}
    >
      {title}
    </NavLink>
  ));

  return (
    <div>
      <AppBar position="static">
        <div className={classes.inner}>
          <Toolbar>
            <IconButton
              edge="start"
              className={classes.menuButton}
              aria-label="toggle mobile menu"
              color="inherit"
              onClick={() => setNavOpen((open) => !open)}
            >
              <MenuIcon />
            </IconButton>
            <Link to="/">
              <img
                src={logo}
                className={classes.logo}
                alt="Crossfeed Icon Navigate Home"
              />
            </Link>
            <div className={classes.mdNav}>{desktopNavItems.slice(0, 3)}</div>
            <div className={classes.lgNav}>
              {desktopNavItems.slice(3, navItems.length)}
            </div>

            <div className={classes.spacing} />
            {userLevel > 0 && (
              <NavLink
                to="/settings"
                activeClassName={classes.activeLink}
                className={clsx(classes.link, classes.userLink)}
                exact={true}
              >
                <UserIcon /> My Account
              </NavLink>
            )}
          </Toolbar>
        </div>
      </AppBar>
      <Drawer
        anchor="left"
        open={navOpen}
        onClose={() => setNavOpen(false)}
        data-testid="mobilenav"
      >
        <List className={classes.mobileNav}>
          {navItems.map(({ title, path }) => (
            <ListItem
              button
              key={title}
              component={NavLink}
              to={path}
              activeClassName={classes.activeMobileLink}
              exact={true}
            >
              {title}
            </ListItem>
          ))}
        </List>
      </Drawer>
    </div>
  );
};
