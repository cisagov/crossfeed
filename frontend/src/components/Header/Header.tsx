import React, { useState } from 'react';
import { NavLink, Link } from 'react-router-dom';
import {
  Header as UsaHeader,
  NavMenuButton,
  Title,
  PrimaryNav
} from '@trussworks/react-uswds';
import logo from './cisa_logo.png';
import classes from './styles.module.scss';
import { useAuthContext } from 'context';

export const Header: React.FC = () => {
  const { currentOrganization, user } = useAuthContext();
  const [mobileExpanded, setMobileExpanded] = useState(false);

  const onExpandMobile: React.MouseEventHandler<HTMLButtonElement> = e => {
    setMobileExpanded(expanded => !expanded);
  };

  const dashboardTitle = 'Dashboard';

  const orgUserNav = [
    <NavLink
      activeClassName="usa-current"
      to="/"
      exact
      key="dashboard"
      className="usa-nav__link"
    >
      <span>{dashboardTitle}</span>
    </NavLink>,
    <NavLink
      activeClassName="usa-current"
      to="/vulnerabilities"
      key="vulnerabilities"
      className="usa-nav__link"
    >
      <span>Vulnerabilities</span>
    </NavLink>,
    <NavLink
      activeClassName="usa-current"
      to="/risk"
      key="risk"
      className="usa-nav__link"
    >
      <span>Risk Summary</span>
    </NavLink>,
    <NavLink
      activeClassName="usa-current"
      to="/organizations"
      key="organizations"
      className="usa-nav__link"
    >
      <span>My Organizations</span>
    </NavLink>,
    <NavLink
      activeClassName="usa-current"
      to="/settings"
      key="settings"
      className="usa-nav__link"
    >
      <span>My Account</span>
    </NavLink>
  ];

  const orgAdminNav = [
    <NavLink
      activeClassName="usa-current"
      to="/"
      exact
      key="dashboard"
      className="usa-nav__link"
    >
      <span>{dashboardTitle}</span>
    </NavLink>,
    <NavLink
      activeClassName="usa-current"
      to="/vulnerabilities"
      key="vulnerabilities"
      className="usa-nav__link"
    >
      <span>Vulnerabilities</span>
    </NavLink>,
    <NavLink
      activeClassName="usa-current"
      to="/risk"
      key="risk"
      className="usa-nav__link"
    >
      <span>Risk Summary</span>
    </NavLink>,
    <NavLink
      activeClassName="usa-current"
      to="/organization"
      key="organization"
      className="usa-nav__link"
    >
      <span>Organization Settings</span>
    </NavLink>,
    <NavLink
      activeClassName="usa-current"
      to="/organizations"
      key="organizations"
      className="usa-nav__link"
    >
      <span>My Organizations</span>
    </NavLink>,
    <NavLink
      activeClassName="usa-current"
      to="/scans"
      key="scans"
      className="usa-nav__link"
    >
      <span>Scans</span>
    </NavLink>,
    <NavLink
      activeClassName="usa-current"
      to="/settings"
      key="settings"
      className="usa-nav__link"
    >
      <span>My Account</span>
    </NavLink>
  ];

  const globalAdminNav = [
    <NavLink
      activeClassName="usa-current"
      to="/"
      exact
      key="dashboard"
      className="usa-nav__link"
    >
      <span>{dashboardTitle}</span>
    </NavLink>,
    <NavLink
      activeClassName="usa-current"
      to="/vulnerabilities"
      key="vulnerabilities"
      className="usa-nav__link"
    >
      <span>Vulnerabilities</span>
    </NavLink>,
    <NavLink
      activeClassName="usa-current"
      to="/risk"
      key="risk"
      className="usa-nav__link"
    >
      <span>Risk Summary</span>
    </NavLink>,
    <NavLink
      activeClassName="usa-current"
      to="/organization"
      key="organization"
      className="usa-nav__link"
    >
      <span>Organization Settings</span>
    </NavLink>,
    <NavLink
      activeClassName="usa-current"
      to="/scans"
      key="scans"
      className="usa-nav__link"
    >
      <span>All Scans</span>
    </NavLink>,
    <NavLink
      activeClassName="usa-current"
      to="/organizations"
      key="organizations"
      className="usa-nav__link"
    >
      <span>Manage Organizations</span>
    </NavLink>,
    <NavLink
      activeClassName="usa-current"
      to="/users"
      key="users"
      className="usa-nav__link"
    >
      <span>Manage Users</span>
    </NavLink>,
    <NavLink
      activeClassName="usa-current"
      to="/settings"
      key="settings"
      className="usa-nav__link"
    >
      <span>My Account</span>
    </NavLink>
  ];

  let nav: JSX.Element[] = [];
  if (user && user.isRegistered) {
    if (user.userType === 'standard') {
      if (currentOrganization?.userIsAdmin) {
        nav = orgAdminNav;
      } else {
        nav = orgUserNav;
      }
    } else {
      nav = globalAdminNav;
    }
  }

  return (
    <UsaHeader basic className={classes.root}>
      <div className="usa-nav-container">
        <div className="usa-navbar">
          <Link to="/">
            <Title>
              <img src={logo} alt="Crossfeed logo" />
            </Title>
          </Link>
          <NavMenuButton label="Menu" onClick={onExpandMobile} />
        </div>

        <PrimaryNav
          mobileExpanded={mobileExpanded}
          onToggleMobileNav={onExpandMobile}
          items={nav}
        />
      </div>
    </UsaHeader>
  );
};
