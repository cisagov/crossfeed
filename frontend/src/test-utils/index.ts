export * from './test-utils';
export * from './user';
export * from './organization';
