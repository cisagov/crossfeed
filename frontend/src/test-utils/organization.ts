import { Organization } from 'types';

export const testOrganization = {
  id: 'd7d6e913-0370-4f43-aebc-6bd727adc70e',
  name: 'test org',
  rootDomains: ['example.com'],
  ipBlocks: [],
  userRoles: [],
  scanTasks: [],
  isPassive: false,
  inviteOnly: false,
  granularScans: []
};
