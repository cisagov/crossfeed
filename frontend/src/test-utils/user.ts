import { AuthUser } from 'context';

export const testUser: AuthUser = {
  id: 'd7d6e913-0370-4f43-aebc-6bd727adc70e',
  createdAt: '2020-08-23T03:36:57.231Z',
  updatedAt: '2020-08-23T03:36:57.231Z',
  firstName: 'John',
  lastName: 'Smith',
  fullName: 'John Smith',
  invitePending: false,
  userType: 'standard',
  email: 'test@crossfeed.gov',
  roles: [],
  dateAcceptedTerms: new Date(),
  acceptedTermsVersion: 'v1-user',
  isRegistered: true
};
