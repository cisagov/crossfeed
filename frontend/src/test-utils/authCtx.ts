import { AuthContextType } from 'context';
import { testUser } from './user';

export const authCtx: AuthContextType = {
  user: testUser,
  token: 'some-test-token-never-verified',
  currentOrganization: null,
  loading: false,
  login: jest.fn(),
  logout: jest.fn(),
  setUser: jest.fn(),
  setOrganization: jest.fn(),
  refreshUser: jest.fn(),
  setLoading: jest.fn(),
  apiGet: jest.fn(),
  apiDelete: jest.fn(),
  apiPatch: jest.fn(),
  apiPost: jest.fn(),
  apiPut: jest.fn(),
  maximumRole: 'user',
  touVersion: 'v1-user',
  userMustSign: false
};
