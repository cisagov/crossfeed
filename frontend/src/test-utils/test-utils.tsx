import React from 'react';
import { render, RenderOptions } from '@testing-library/react';
import { mocked } from 'ts-jest/utils';
import { MemoryRouter } from 'react-router-dom';
import { AuthContext, AuthContextType } from '../context/AuthContext';
import { CFThemeProvider } from 'context';
import { authCtx } from './authCtx';

interface CustomRenderOptions extends RenderOptions {
  initialHistory?: string[];
  authContext?: Partial<AuthContextType>;
}

const customRender = (ui: any, options: CustomRenderOptions = {}) => {
  const { initialHistory, authContext, ...rest } = options;

  // Provide any context that the components may be expecting
  const Wrapper: React.FC = ({ children }) => (
    <CFThemeProvider>
      <MemoryRouter initialEntries={initialHistory}>
        <AuthContext.Provider
          value={{
            ...authCtx,
            ...authContext
          }}
        >
          {children}
        </AuthContext.Provider>
      </MemoryRouter>
    </CFThemeProvider>
  );
  return render(ui, { wrapper: Wrapper, ...rest });
};

// re-export everything
export * from '@testing-library/react';

// override render method
export { customRender as render, mocked };
