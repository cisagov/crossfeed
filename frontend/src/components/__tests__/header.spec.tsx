import React from 'react';
import { render, fireEvent, testUser, testOrganization } from 'test-utils';
import { Header } from '../Header';
import { wait, getByTestId } from '@testing-library/react';

jest.mock('@elastic/react-search-ui', () => ({
  withSearch: () => (comp: any) => comp
}));

describe('Header component', () => {
  it('matches snapshot', () => {
    const { asFragment } = render(<Header />);
    expect(asFragment()).toMatchSnapshot();
  });

  it('can expand drawer', async () => {
    const { getByLabelText, getByTestId, queryByTestId } = render(<Header />);
    expect(queryByTestId('mobilenav')).not.toBeInTheDocument();
    expect(getByLabelText('toggle mobile menu')).toBeInTheDocument();
    fireEvent.click(getByLabelText('toggle mobile menu'));
    await wait(() => {
      expect(getByTestId('mobilenav')).toBeInTheDocument();
    });
  });

  it('shows no links for unauthenticated user', () => {
    const { queryByText } = render(<Header />, {
      authContext: {
        user: { ...testUser, userType: 'standard', isRegistered: false },
        currentOrganization: { ...testOrganization, userIsAdmin: false }
      }
    });
    [
      'Vulnerabilities',
      'Risk Summary',
      'My Organizations',
      'Manage Organizations',
      'Scans',
      'Manage Users',
      'My Account'
    ].forEach((expected) => {
      expect(queryByText(expected)).not.toBeInTheDocument();
    });
  });

  it('shows correct links for ORG_USER', () => {
    const { getByText, queryByText } = render(<Header />, {
      authContext: {
        user: { ...testUser, userType: 'standard', isRegistered: true },
        currentOrganization: { ...testOrganization, userIsAdmin: false }
      }
    });
    [
      'Overview',
      'Inventory',
      'My Organizations',
      'My Account',
      'My Settings',
      'Logout'
    ].forEach((expected) => {
      expect(getByText(expected)).toBeInTheDocument();
    });
    ['Manage Organizations', 'Scans', 'Manage Users'].forEach((notExpected) => {
      expect(queryByText(notExpected)).not.toBeInTheDocument();
    });
  });

  it('shows correct links for ORG_ADMIN', () => {
    const { getByText, queryByText } = render(<Header />, {
      authContext: {
        user: { ...testUser, userType: 'standard', isRegistered: true },
        currentOrganization: { ...testOrganization, userIsAdmin: true }
      }
    });
    [
      'Overview',
      'Inventory',
      'Organization Settings',
      'My Organizations',
      'My Account',
      'My Settings',
      'Logout'
    ].forEach((expected) => {
      expect(getByText(expected)).toBeInTheDocument();
    });
    ['Manage Organizations', 'Manage Users'].forEach((notExpected) => {
      expect(queryByText(notExpected)).not.toBeInTheDocument();
    });
  });

  it('shows correct links for GLOBAL_ADMIN', () => {
    const { getByText, queryByText } = render(<Header />, {
      authContext: {
        user: { ...testUser, userType: 'globalAdmin', isRegistered: true },
        currentOrganization: { ...testOrganization, userIsAdmin: true }
      }
    });
    [
      'Overview',
      'Inventory',
      'Scans',
      'Organization Settings',
      'Manage Organizations',
      'Manage Users',
      'My Account',
      'My Settings',
      'Logout'
    ].forEach((expected) => {
      expect(getByText(expected)).toBeInTheDocument();
    });
    ['My Organizations'].forEach((notExpected) => {
      expect(queryByText(notExpected)).not.toBeInTheDocument();
    });
  });
});
