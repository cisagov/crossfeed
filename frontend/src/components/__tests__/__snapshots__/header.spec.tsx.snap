// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Header component matches snapshot 1`] = `
<DocumentFragment>
  <div>
    <header
      class="MuiPaper-root MuiAppBar-root MuiAppBar-positionStatic MuiAppBar-colorPrimary MuiPaper-elevation0"
    >
      <div
        class="makeStyles-inner-1"
      >
        <div
          class="MuiToolbar-root MuiToolbar-regular MuiToolbar-gutters"
        >
          <button
            aria-label="toggle mobile menu"
            class="MuiButtonBase-root MuiIconButton-root makeStyles-menuButton-2 MuiIconButton-colorInherit MuiIconButton-edgeStart"
            tabindex="0"
            type="button"
          >
            <span
              class="MuiIconButton-label"
            >
              <svg
                aria-hidden="true"
                class="MuiSvgIcon-root"
                focusable="false"
                viewBox="0 0 24 24"
              >
                <path
                  d="M3 18h18v-2H3v2zm0-5h18v-2H3v2zm0-7v2h18V6H3z"
                />
              </svg>
            </span>
            <span
              class="MuiTouchRipple-root"
            />
          </button>
          <a
            href="/"
          >
            <img
              alt="Crossfeed Icon Navigate Home"
              class="makeStyles-logo-3"
              src="cisa_logo.png"
            />
          </a>
          <div
            class="makeStyles-mdNav-9"
          >
            <a
              aria-current="page"
              class="makeStyles-link-16 makeStyles-activeLink-14"
              href="/"
              style="outline: none;"
            >
              Overview
            </a>
            <a
              class="makeStyles-link-16"
              href="/inventory"
              style="outline: none;"
            >
              Inventory
            </a>
          </div>
          <div
            class="makeStyles-lgNav-10"
          />
          <div
            class="makeStyles-spacing-4"
          />
          <div
            class="makeStyles-wrapper-18"
          >
            <div
              class="makeStyles-inner-19"
            >
              <svg
                aria-hidden="true"
                class="MuiSvgIcon-root makeStyles-icon-21"
                focusable="false"
                viewBox="0 0 24 24"
              >
                <path
                  d="M15.5 14h-.79l-.28-.27C15.41 12.59 16 11.11 16 9.5 16 5.91 13.09 3 9.5 3S3 5.91 3 9.5 5.91 16 9.5 16c1.61 0 3.09-.59 4.23-1.57l.27.28v.79l5 4.99L20.49 19l-4.99-5zm-6 0C7.01 14 5 11.99 5 9.5S7.01 5 9.5 5 14 7.01 14 9.5 11.99 14 9.5 14z"
                />
              </svg>
              <form>
                <input
                  class="makeStyles-inp-20"
                  placeholder="Search for a domain, vuln type, port, service, IP"
                  value=""
                />
              </form>
            </div>
          </div>
          <a
            aria-current="page"
            class="makeStyles-link-16 makeStyles-link-16"
            href="/"
            style="outline: none;"
          >
            <div
              class="makeStyles-userLink-8"
            >
              <svg
                aria-hidden="true"
                class="MuiSvgIcon-root"
                focusable="false"
                viewBox="0 0 24 24"
              >
                <path
                  d="M12 2C6.48 2 2 6.48 2 12s4.48 10 10 10 10-4.48 10-10S17.52 2 12 2zm0 3c1.66 0 3 1.34 3 3s-1.34 3-3 3-3-1.34-3-3 1.34-3 3-3zm0 14.2c-2.5 0-4.71-1.28-6-3.22.03-1.99 4-3.08 6-3.08 1.99 0 5.97 1.09 6 3.08-1.29 1.94-3.5 3.22-6 3.22z"
                />
              </svg>
               My Account 
              <svg
                aria-hidden="true"
                class="MuiSvgIcon-root"
                focusable="false"
                viewBox="0 0 24 24"
              >
                <path
                  d="M7 10l5 5 5-5z"
                />
              </svg>
            </div>
          </a>
        </div>
      </div>
    </header>
  </div>
</DocumentFragment>
`;
